SIPS/sip-20.md

```diff
@@ -1,191 +1,313 @@
- *
- * @property type - The type of the external data source.
- * @property data - The data received from the external data source.
- */
-type ExternalData = WebSocketData;
+export type WebSocketMessageData =
+  | WebSocketTextMessage
+  | WebSocketBinaryMessage;
 
-type OnExternalDataHandlerArgs = ExternalData & {
-  source: string;
+export type WebSocketMessage = {
+  type: "message";
+  id: string;
+  origin: string;
+  data: WebSocketMessageData;
 };
+```
 
-type OnExternalDataHandler = (args: OnExternalDataHandlerArgs) => Promise<void>;
+`type` - The type of the WebSocket event, which is `"message"` for this event type.
+
+`id` - The unique identifier of the WebSocket connection associated with the event.
+
+`origin` - The origin of the WebSocket event.
+
+`data` - The data received in the event. This can be either a text message (as a string) or binary data (as an array of numbers). The `type` property indicates whether the data is text or binary.
+
+##### Connection opened event
+
+This event is triggered when a WebSocket connection is successfully opened. It provides the unique identifier of the connection and the origin of the event.
+
+```typescript
+export type WebSocketOpenEvent = {
+  type: "open";
+  id: string;
+  origin: string;
+};
 ```
+`type` - The type of the WebSocket event, which is `"open"` for this event type.
+
+`id` - The unique identifier of the WebSocket connection associated with the event.
 
-#### Example
+`origin` - The origin of the WebSocket connection.
+
+##### Connection closed event
+
+This event is triggered when a WebSocket connection is closed. It provides the unique identifier of the connection and the origin of the event.
 
 ```typescript
-import { OnExternalDataHandler } from '@metamask/snaps-types';
-import { assert } from '@metamask/utils';
+export type WebSocketCloseEvent = {
+  type: "close";
+  id: string;
+  origin: string;
+  code: number;
+  reason: string;
+};
+```
+
+`type` - The type of the WebSocket event, which is `"close"` for this event type.
+
+`id` - The unique identifier of the WebSocket connection associated with the event.
+
+`origin` - The origin of the WebSocket connection.
+
+`code` - The numeric code indicating the reason for the closure of the WebSocket connection. This is a standard WebSocket close code.
+
+`reason` - A string providing a human-readable explanation of why the WebSocket connection was closed.
 
-export const onExternalData: OnExternalDataHandler = ({ type, data, source }) => {
-  assert(type === 'websocket'); // `data` is inferred as `WebSocketData`.
-  assert(data.type === 'text'); // `message` is inferred as `string`.
+##### Error event
 
-  // Now the Snap can do whatever is needed with the message.
-  const json = JSON.parse(data.message);
+This event is triggered when an error occurs with a WebSocket connection. It provides the unique identifier of the connection, and the origin of the event.
+
+```typescript
+export type WebSocketErrorEvent = {
```

Comment:
This should probably include the error? Or does the WebSocket API not give an actual error message?

Reply:
The `error` event don't give any error message, the event type for errors is a generic `Event`. You get a code and a reason in the `close` event, that's why I just added these properties to the close event

Reply:
https://stackoverflow.com/a/18804298

Reply:
Do we need to emit this event at all then?

Reply:
It might not be useful if it provides no information at all 😅

Reply:
Hmm, I was assuming an error would always close, but maybe not 🤷‍♂️

Reply:
This would allow us to be close to the og. WS